### PR TITLE
Merge a rebased and cleaned up version of the feature-gptp-avnu-automotive-profile branch

### LINKED ISSUE #360: gPTP - auto profile, remove debugout..hpp
Since the automotive profile feature branch adds (and uses) gptp_log.hpp, debugout.hpp should be removed.

### PR DESCRIPTION
**Pull request by pinealservo**
The original implementation work by @kencarlino was done against a much earlier version of the tree, so there were significant places where automatic merging failed. In this branch, I have split the original changes into a single merge of the automotive profile support followed by a number of cleanup patches that apply some new functionality introduced more evenly across the code or which fix problems introduced related to building in Windows.

Additionally, there is a fix for issue #388 (of which #401 is a duplicate) included since I discovered it in the process of testing this branch.  This also resolves #360 as part of the logging system consolidation.

There were some whitespace issues regarding mixed tabs/spaces in the same files that were cleaned up during the merge; you may want to view ignoring whitespace changes for the review.

### CODE REVIEW

**Comment by pinealservo** [comment]
Oh, the joys of an ancient compiler.  Travis's gcc doesn't like --std=c++11.

**Comment by andrew-elder** [comment]
As far as I can tell this branch behaves the same as open-avb-next in Windows 10 (with I217-LM at least). I assume Linux operation with an i210 has been confirmed by Levi?

**Comment by andrew-elder** [comment]
Can some folks commit to reviewing the patch?

**Comment by andrew-elder** [comment]
Log output in Windows has extra spaces....
```
ERROR    : GPTP [09:08:17:597] Discarding invalid message
STATUS   : GPTP [09:08:20:202] *** Announce Timeout Expired - Becoming Master

STATUS   : GPTP [09:08:20:202] New Grandmaster "F8:B1:56:FF:FE:E4:0E:DA" (previous "00:00:00:00:00:00:00:00")

STATUS   : GPTP [09:08:37:544] Switching to Slave

STATUS   : GPTP [09:08:37:544] New Grandmaster "00:04:96:FF:FE:8B:C9:3F" (previous "F8:B1:56:FF:FE:E4:0E:DA")

STATUS   : GPTP [09:09:49:774] Switching to Master

STATUS   : GPTP [09:09:49:774] Switching to Slave
```

**Comment by davidcemin** [comment]
Its a big patch. I can review it by tomorrow morning.

**Comment by RikusW** [comment]
#401 seems to be fixed, I fetched and compiled pinealservo:task-merge-automotive-pullrequest

**Comment by andrew-elder** [comment]
Thanks @RikusW. That is useful information.

**Comment by pinealservo** [comment]
> Log output in Windows has extra spaces....

I believe those are actually literal newlines in the log messages.  They are there in some of them, and not there in others, and I didn't change any of them.  I can add a cleanup of those to this pull request, or we can do it after; what's your preference?

**Comment by andrew-elder** [comment]
Let's plan to cleanup extra newlines in the log messages after this pull request.

File: daemons/gptp/README_AVNU_AP.txt [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+- Brief summary to give a sense of scope of updates needed for Windows
+	- MakeFile 
+		- Updates for new files
+	- daemon_cl.cpp
+		- New command line options
+		- Port init via struct rather than separate parameters for the constructor
+		- Changed state save and restore (persistence)
+		- Automotive profile configuration setup
+		- Updates to signal handling
+		- Logging changes
```
Comment (andrew-elder) [inline daemons/gptp/README_AVNU_AP.txt]: We should delete "Logging changes" from here and the following comments as this pull request addresses required changes.

File: daemons/gptp/common/avbts_persist.hpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ * Generic interface for Simple Persistence for gPTP values
+ */
+class GPTPPersist {
+public:
+	/**
+	 * @brief  Initializes the GPTP_PERSIST
+	 * @param persistID string identifer of the persistence storage. For example a file name
+	 * @return True on success otherwise False
+	 */
+    virtual bool initStorage(const char *persistID) = 0;
```
Comment (andrew-elder) [inline daemons/gptp/common/avbts_persist.hpp]: Indent?

File: daemons/gptp/common/avbts_port.hpp [hunk 3]
```diff
@@ -202,11 +207,95 @@ class PortIdentity {
 typedef std::map < PortIdentity, LinkLayerAddress > IdentityMap_t;
 
 /**
+ * Structure for initializing the IEEE1588 class
+ */
+typedef struct {
+	/* clock IEEE1588Clock instance */
```
Comment (andrew-elder) [inline daemons/gptp/common/avbts_port.hpp]: Comment is not Doxygen compatible .....

File: daemons/gptp/windows/daemon_cl/platform.cpp [hunk 4]
```diff
@@ -63,3 +64,8 @@ uint64_t PLAT_ntohll(uint64_t x)
 {
 	return( (ntohl(1) == 1) ? x : ((uint64_t)ntohl((x) & 0xFFFFFFFF) << 32) | ntohl((x) >> 32) );
 }
+
+errno_t PLAT_localtime(const time_t * inTime, struct tm * outTm)
+{
+	return localtime_s(outTm, inTime);
+}
```
Comment (andrew-elder) [inline daemons/gptp/windows/daemon_cl/platform.cpp]: No newline

**Comment by andrew-elder** [comment]
I've read the pull request and only have a few minor comments. That said, the pull request is really too large to review by simple visual inspection. I'm hoping some other folks can try running it as well.

File: daemons/gptp/README_AVNU_AP.txt (hunk 1 above)
Comment (pinealservo) [inline daemons/gptp/README_AVNU_AP.txt]: Now that I look through it again, I also did the "Port init via struct..." one.  Removing those lines.

File: daemons/gptp/common/avbts_port.hpp (hunk 3 above)
Comment (pinealservo) [inline daemons/gptp/common/avbts_port.hpp]: Hmm, I didn't catch that because I usually have JAVADOC_AUTOBRIEF turned on in the Doxygen config (and it looks like the Harman-STC folks do as well based on the avtp-pipeline Doxygen config), in which case the first line of text becomes the 'brief' description automatically.  We don't have that turned on for gptp, though, so you're correct that it won't get picked up as a Doxygen comment.  I'll fix them.

Comment (pinealservo) [inline daemons/gptp/common/avbts_port.hpp]: It turns out that this is an *incredibly* common thing. I did a regex search for Doxygen-style comments without any Doxygen tags and fixed them all, but it ended up touching most of the .h/.hpp files in the gptp daemon source. It may be best to make a separate pull request for it afterward.

**Comment by pinealservo** [comment]
I will be on vacation for the rest of the week; I'll respond to any further review comments on Monday.

File: daemons/gptp/common/avbap_message.hpp [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+#define AP_TEST_STATUS_COUNTER_29(x) x + 148					/*!< Counter 29 */
+#define AP_TEST_STATUS_COUNTER_30(x) x + 152					/*!< Counter 30 */
+#define AP_TEST_STATUS_COUNTER_31(x) x + 156					/*!< Counter 31 */
+
+
+/**
+ * Automotive Profile Test Status Station State
+ */
+typedef enum {
+	STATION_STATE_RESERVED,
```
Comment (davidcemin) [inline daemons/gptp/common/avbap_message.hpp]: should we add a doxygen comment for each one of the states?

File: daemons/gptp/common/avbts_message.hpp [hunk 6]
```diff
@@ -689,71 +694,80 @@ class FollowUpTLV {
...
-
-    /**
-     * @brief  Gets the scaledLastGmPhaseChange private member value
-     * @return scaledLastGmPhaseChange value
-     */
-    scaledNs getScaledLastGmPhaseChange(void)
-    {
-        return scaledLastGmPhaseChange;
-    }
+	/**
```
Comment (davidcemin) [inline daemons/gptp/common/avbts_message.hpp]: It looks like you changed the indentation on this file. Was this intentional ?

File: daemons/gptp/common/avbts_message.hpp [hunk 7]
```diff
@@ -1010,4 +1024,160 @@ class PTPMessagePathDelayRespFollowUp:public PTPMessageCommon {
...
+ * Provides a class for building a PTP signalling message
+ */
+class PTPMessageSignalling:public PTPMessageCommon {
+private:
+	int8_t targetPortIdentify;
+	SignallingTLV tlv;
+
+	PTPMessageSignalling(void);
+public:
+	static const int8_t sigMsgInterval_Initial =  126;
```
Comment (davidcemin) [inline daemons/gptp/common/avbts_message.hpp]: These constants are public, so shouldn't we document them as well?

File: daemons/gptp/common/avbts_port.hpp (hunk 3 above)
Comment (davidcemin) [inline daemons/gptp/common/avbts_port.hpp]: I agree on separating the pull request for doxygen related things.

File: daemons/gptp/common/gptp_log.hpp [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+#include <stdarg.h>
+#include <time.h>
+
+#define GPTP_LOG_CRITICAL_ON		1
+#define GPTP_LOG_ERROR_ON			1
+#define GPTP_LOG_EXCEPTION_ON	1
+#define GPTP_LOG_WARNING_ON		1
+#define GPTP_LOG_INFO_ON				1
+#define GPTP_LOG_STATUS_ON			1
+//#define GPTP_LOG_DEBUG_ON			1
```
Comment (davidcemin) [inline daemons/gptp/common/gptp_log.hpp]: Is there a way to change the log level without having to recompile the gptp code?

File: daemons/gptp/common/ieee1588clock.cpp [hunk 9]
```diff
@@ -428,14 +420,14 @@ bool IEEE1588Clock::isBetterThan(PTPMessageAnnounce * msg)
 	msg->getGrandmasterIdentity((char *)that1 + 6);
 
 #if 0
-	XPTPD_PRINTF("(Clk)Us: ");
+	GPTP_LOG_DEBUG("(Clk)Us: ");
```
Comment (davidcemin) [inline daemons/gptp/common/ieee1588clock.cpp]: Do we need this code removed with if 0? It looks like the GPTP_LOG_DEBUG will only be printing this out when the DEBUG define is set.

File: daemons/gptp/common/ieee1588port.cpp [hunk 10]
```diff
@@ -79,47 +82,73 @@ IEEE1588Port::~IEEE1588Port()
...
+		if (operLogPdelayReqInterval == LOG2_INTERVAL_INVALID)
+			operLogPdelayReqInterval = 0;      // 1 second
+		if (operLogSyncInterval == LOG2_INTERVAL_INVALID)
+			operLogSyncInterval = 0;           // 1 second
+	}
+	else {
+		asCapable = false;
+
+		if (initialLogSyncInterval == LOG2_INTERVAL_INVALID)
+			initialLogSyncInterval = -3;       // 125 ms
```
Comment (davidcemin) [inline daemons/gptp/common/ieee1588port.cpp]: I might be missing something here, but dont we have a constant somewhere for logsyncinterval (and the others as well) ? Then we wouldnt have to hard code these values here (-5, -3 etc).

File: daemons/gptp/common/ieee1588port.cpp [hunk 11]
```diff
@@ -135,25 +164,37 @@ IEEE1588Port::IEEE1588Port
...
+	this->condition_factory = portInit->condition_factory;
+	this->lock_factory = portInit->lock_factory;
 
 	pdelay_count = 0;
 	sync_count = 0;
 	memset(link_delay, 0, sizeof(link_delay));
+
+	if (testMode) {
+		if (isGM) {
+			avbSyncState = 1;
```
Comment (davidcemin) [inline daemons/gptp/common/ieee1588port.cpp]: What does the avbSyncState values mean ? I gather that 1 means GM and 2 means not GM. Should we create a definition for these values?

File: daemons/gptp/common/ieee1588port.cpp [hunk 12]
```diff
@@ -477,427 +568,528 @@ void IEEE1588Port::processEvent(Event e)
...
+				PTPMessageAnnounce *EBest = NULL;
+				char EBestClockIdentity[PTP_CLOCK_IDENTITY_LENGTH];
+
+				IEEE1588Port **ports;
+				clock->getPortList(number_ports, ports);
+
+				/* Find EBest for all ports */
+				j = 0;
+				for (int i = 0; i < number_ports; ++i) {
+					while (ports[j] == NULL)
```
Comment (davidcemin) [inline daemons/gptp/common/ieee1588port.cpp]: does getPortList guarantee that ports is always a valid pointer and that number_ports is always within the range of the pointer? (there is no error check on return of clock->getPortList)

File: daemons/gptp/common/ptp_message.cpp [hunk 13]
```diff
@@ -136,30 +136,29 @@ PTPMessageCommon *buildPTPMessage
 		if (ts_good != GPTP_EC_SUCCESS) {
 			char msg[HWTIMESTAMPER_EXTENDED_MESSAGE_SIZE];
 			port->getExtendedError(msg);
-			XPTPD_ERROR
+			GPTP_LOG_ERROR
 			    ("*** Received an event packet but cannot retrieve timestamp, discarding. messageType=%u,error=%d\n%s",
 			     messageType, ts_good, msg);
 			//_exit(-1);
```
Comment (davidcemin) [inline daemons/gptp/common/ptp_message.cpp]: I dont this code was already here, but it looks like dead code to me. Should we remove this line?

File: daemons/gptp/common/ieee1588port.cpp [hunk 14]
```diff
@@ -989,43 +1213,43 @@ void IEEE1588Port::getDeviceTime
...
-  clock->deleteEventTimer( this, SYNC_INTERVAL_TIMEOUT_EXPIRES );
+	clock->deleteEventTimer( this, ANNOUNCE_INTERVAL_TIMEOUT_EXPIRES );
+	clock->deleteEventTimer( this, SYNC_INTERVAL_TIMEOUT_EXPIRES );
 
-  port_state = PTP_SLAVE;
+	port_state = PTP_SLAVE;
 
-  /*clock->addEventTimer
-	  ( this, SYNC_RECEIPT_TIMEOUT_EXPIRES,
+	/*clock->addEventTimer
```
Comment (davidcemin) [inline daemons/gptp/common/ieee1588port.cpp]: Should we remove dead code?

File: daemons/gptp/common/ptp_message.cpp [hunk 15]
```diff
@@ -465,10 +464,27 @@ PTPMessageCommon *buildPTPMessage
 			msg = annc;
 		}
 		break;
+
+	case SIGNALLING_MESSAGE:
+		{
+			PTPMessageSignalling *signallingMsg = new PTPMessageSignalling();
```
Comment (davidcemin) [inline daemons/gptp/common/ptp_message.cpp]: sometimes we check for malloc (new) returns, sometimes we dont. Is there a best practice item for this within open-avb?

File: daemons/gptp/common/ptp_message.cpp [hunk 16]
```diff
@@ -616,14 +632,14 @@ bool PTPMessageAnnounce::isBetterThan(PTPMessageAnnounce * msg)
 	msg->getGrandmasterIdentity((char *)that1 + 6);
 
 #if 0
-	XPTPD_PRINTF("Us: ");
+	GPTP_LOG_VERBOSE("Us: ");
```
Comment (davidcemin) [inline daemons/gptp/common/ptp_message.cpp]: Do we need the if 0? Looks like log_verbose will only print if defined

File: daemons/gptp/common/ptp_message.cpp [hunk 17]
```diff
@@ -1648,3 +1693,130 @@ void PTPMessagePathDelayRespFollowUp::setRequestingPortIdentity
...
+	tlv.setAnnounceInterval(announceInterval);
+}
+
+void PTPMessageSignalling::sendPort(IEEE1588Port * port, PortIdentity * destIdentity)
+{
+	uint8_t buf_t[256];
+	uint8_t *buf_ptr = buf_t + port->getPayloadOffset();
+	unsigned char tspec_msg_t = 0x0;
+
+	memset(buf_t, 0, 256);
```
Comment (davidcemin) [inline daemons/gptp/common/ptp_message.cpp]: sizeof(buf_t) instead of 256 ?

**Comment by davidcemin** [comment]
It is a big review, and as stated on the pull request, there is a bunch of white space changes, which makes the review look bigger than it actually is. 
Didnt we review this code before on another automotive profile pull request? if so, we might be good to integrate this as is. 
My comments are minor and should not block the integration of this pull request.

**Comment by andrew-elder** [comment]
@christopher-s-hall - are you ok if we merge this pull request as is and do cleanups separately? I think it is probably the best idea.

**Comment by andrew-elder** [comment]
Consensus is to merge as is. Merging it...
